### PR TITLE
htop: replace individual option with 'settings'

### PR DESCRIPTION
**Pull request by bjpbakker**
### Description

Add some meters supported in htop but missing from the configuration.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Comment by berbiche** [comment]
Hi, thanks for the contribution.

How can I verify these values?

**Comment by bjpbakker** [comment]
These values come from `htop`s builtin setup that I used to tweak the left and right meters.

You could do a cross check with the [htop](https://github.com/htop-dev/htop). The individual `*Meter.c` files define the meter names used in config. Some grepping lead to me the following possible meters. I haven't checked all of them for being supported, but happy to do that and add any missing ones to this PR if you like.

```
AllCPUs
AllCPUs2
AllCPUs4
AllCPUs8
Battery
Blank
Clock
CPU
Date
DateTime
DiskIO
Hostname
HugePages
LeftCPUs
LeftCPUs2
LeftCPUs4
LeftCPUs8
Load
LoadAverage
Memory
NetworkIO
PressureStallCPUSome
PressureStallIOFull
PressureStallIOSome
PressureStallMemoryFull
PressureStallMemorySome
RightCPUs
RightCPUs2
RightCPUs4
RightCPUs8
SELinux
Swap
System
Systemd
Tasks
Uptime
ZFSARC
ZFSCARC
Zram
```

**Comment by teto** [comment]
It looks overkill to me to provide this level of detail for configuration (a losing battle) but now that we've started doing it, maybe you should add all of them ^^

**Comment by rycee** [comment]
I would suggest deprecating pretty much all the options in the htop module and instead add a plain settings option. Maybe adding the fields as constants in `config.lib.htop.fields` so that people can refer to names instead of numbers.

**Comment by bjpbakker** [comment]
> It looks overkill to me to provide this level of detail for configuration (a losing battle)

> I would suggest deprecating pretty much all the options in the htop module and instead add a plain settings option.

Agreed, that should be the way to go on the long run.

Either way, I've added the missing ones for now (checked against htop@2328e52).

**Comment by teto** [comment]
would be nice to do the depreciation before the nixos 21.05 release.

**Comment by bjpbakker** [comment]
I can definitely deprecate all options and replace them with a single `settings` that you basically paste the htoprc contents into.

One of the benefits of doing that, is that people can use setup in `htop` itself to build their configuration and then paste it in their home-manager config. That works if all specific settings are dropped from nix configuration.

I aim to open a new PR with those changes somewhere next week.

**Comment by teto** [comment]
wouldn't it be possible to update this one instead ?

**Comment by bjpbakker** [comment]
> wouldn't it be possible to update this one instead ?

Sure, it's just that the scope of this has changed a bit, but I can also push it to this PR.

**Comment by bjpbakker** [comment]
I've deprecated all options and introduced `settings` for setting htop configuration values from Nix config. Some `lib.htop` predefined fields and modes, and functions for meters are provided for convenience.

Hope this helps.

@teto sorry it took a while, was quite busy the last couple weeks. Let me know any feedback, happy to address it :)

File: modules/programs/htop.nix [hunk 1]
```diff
@@ -368,49 +537,60 @@ in {
...
+      description = ''
+        Deprecated. Please use programs.htop.settings.vim_mode instead.
+
+        Vim key bindings.
+      '';
     };
   };
 
   config = mkIf cfg.enable {
+    warnings = optional (cfg.settings == { }) ''
```
Comment (teto) [inline modules/programs/htop.nix:549]: This looks like a dubious check, maybe in the default settings' prefix cfg.X with `lib.warn "warning" cfg.X` isntead. Also maybe add a release note. I dont know when the deprecation will be removed. Maybe just after the release of nixos.

Comment (bjpbakker) [reply modules/programs/htop.nix]: Agreed, this one is ugly.

A few days ago I already pushed an added release note. Now also a commit to redo the configuration a bit. Moved defaults into the `settings` attrset and show a deprecation warning on any (deprecated) option that is explicitly specified in user config. Messages are much more clear now and should properly inform what to do, without relying on reading releases notes too much :)

If this PR is good to go, I'll draft up another one to remove all deprecated things (to merge after 21.05 release, or when suitable).

**Comment by bjpbakker** [comment]
So, is this one good to go now, with release notes updated and better deprecation warnings?

/cc: @teto

**Comment by teto** [comment]
can you fix the tests please ?

**Comment by bjpbakker** [comment]
> can you fix the tests please ?

@teto sorry about that, unfortunate timing of picking previous master. I did run the htop tests which were fine, didn't run the entire suite myself again. Fixed by rebasing on top of current master and checking the full suite locally.

@rycee since you are the current code owner, would you like me to put my name down as new code owner of this module?

**Comment by teto** [comment]
I think @rycee would be happy to cover less ground so I would favor you becming the new owner :)

**Comment by bjpbakker** [comment]
@teto I now see more issues in the install target of the CI build; fixing them.

**Comment by rvl** [comment]
With the `config.lib.leftMeters` function, how would a user control the ordering of meters in the column?

**Pull request merged**